src/internals/strings/impl/length.ts

```diff
@@ -1,66 +1,42 @@
 // Original impl https://gist.github.com/sno2/7dac868ec6d11abb75250ce5e2b36041
 
 import { Add } from "../../numbers/impl/addition";
-import { Mul } from "../../numbers/impl/multiply";
+import { StringIterator as StrIter } from "./utils";
 
-type TCache = [string, 0[]];
+// Implementation of the following algorithm:
+//
+// function length(s, acc) {
+//   if (s === "") {
+//       return 0;
+//   } else if (s.indexOf(acc) === 0) {
+//       return acc.length + length(s.slice(acc.length), acc + acc);
+//   } else {
+//       return length(s, acc.slice(acc.length / 2));
+//   }
+// }
 
 type LengthUp<
-  T extends string,
-  $Acc extends 0[] = [],
-  $Cache extends TCache[] = [[`$${string}`, [0]]]
-> =
-  //
-  $Cache extends [infer C extends TCache, ...infer $RestCache extends TCache[]]
-    ? (
-        `${C[0]}${C[0]}_` extends `$${string}$${infer $After}`
-          ? `${C[0]}${$After}` extends `${infer $Before}_`
-            ? $Before
-            : never
-          : never
-      ) extends infer $DoubleC extends string
-      ? `$${T}` extends `${$DoubleC}${infer $Rest}`
-        ? $Cache["length"] extends 12 // 2^13 is the last block size within the complexity limit
-          ? LengthDownSlow<
-              $Rest,
-              Add<$Acc["length"], Mul<C[1]["length"], 2>>,
-              [[$DoubleC, [...C[1], ...C[1]]], ...$Cache]
-            >
-          : LengthUp<
-              $Rest,
-              [...$Acc, ...C[1], ...C[1]],
-              [[$DoubleC, [...C[1], ...C[1]]], ...$Cache]
-            >
-        : `$${T}` extends `${C[0]}${infer $Rest}`
-        ? LengthUp<$Rest, [...$Acc, ...C[1]], $Cache>
-        : LengthDown<T, $Acc, $RestCache>
-      : never
-    : $Acc["length"];
+  Str extends string,
+  $Length extends number | bigint = 0,
+  It extends StrIter.Iterator = StrIter.Init
+> = StrIter.Double<It> extends infer $DoubleIt extends StrIter.Iterator
+  ? StrIter.CutAt<Str, $DoubleIt> extends `${infer $Rest}`
```

Comment:
isn't this a `Slice<0, _>`?

Reply:
Technically yes. 
But we certainly don't wan't to implement it with slice that is implemented with `split`, `Take` and `Drop` and `Join`. 
Slice should be implemented with this StringIterator abstraction we just made.